### PR TITLE
Bach BQ: SeriesJson slicing with dicts

### PR DESCRIPTION
**Pull request by KathiaBarahona**
Followed this approach: https://docs.google.com/document/d/1yPt4ML5liYgqiRbbJmhW_j3Co5PkSA0539nKocb25xg

### CODE REVIEW

File: bach/tests/functional/bach/test_series_json.py [hunk 1]
```diff
@@ -261,34 +261,39 @@ def test_json_getitem_query(pg_engine, dtype):
             [4, []]
         ]
     )
+
+@pytest.mark.skip_postgres
+def test_json_same_start_end_slicing(engine, dtype):
```
Comment (KathiaBarahona) [inline bach/tests/functional/bach/test_series_json.py:266]: https://github.com/objectiv/objectiv-analytics/issues/853

File: bach/bach/series/series_json.py [hunk 2]
```diff
@@ -479,6 +488,36 @@ def _get_slice_partial_expr(self, value: Optional[int], is_start: bool) -> Expre
...
+
+        :param filtering_slice: dictionary containing the keys and values to use on filter.
+        :param is_start: whether value is the slice.start (True) or slice.stop (False) value. If true,
+            the first element containing the values is returned, otherwise the last.
+        :return: Expression that gets the position of the element based on the slicing filters.
+        """
+        if not isinstance(filtering_slice, dict):
+            raise TypeError(f'key should be a dict, actual type: {type(filtering_slice)}')
+
+        # this way we can reuse code from the accesor without duplicating
```
Comment (thijs-obj) [inline bach/bach/series/series_json.py:505]: clever

Comment (jansentom) [inline bach/bach/series/series_json.py:505]: I was a bit confused about what `element` mean, until I discovered it in the query template below. Maybe this helps for others suffering from local blindness :)

```suggestion
        # this way we can reuse code from the accesor without duplicating.
        # 'element' will filled by iterating over the json array in the query below.
```

File: bach/tests/functional/bach/test_series_json.py (hunk 1 above)
Comment (jansentom) [reply bach/tests/functional/bach/test_series_json.py:266]: Hmm, does @hendrik-obj rely on this behaviour?

Comment (jansentom) [reply bach/tests/functional/bach/test_series_json.py:266]: The pydoc in the accessor also states the "broken" behaviour if I understand you correctly. Should we follow `.loc` behaviour here or invent our own? WDYT?

**Review by jansentom — COMMENTED** [review]
Very simple solution actually and useful to have! Let's check what to do with in/exclusive end of slice behaviour. @hendrik-obj @vard-obj Please shine your light as well, thanks!

**Review by thijs-obj — APPROVED** [review]
lgtm

I do think that we should not 'abuse' the slice operation for this. Don't think mypy likes that either. But that's a separate issue that we need to discuss with everyone involved.

File: bach/bach/series/series_json.py (hunk 2 above)
Comment (KathiaBarahona) [reply bach/bach/series/series_json.py]: Done in https://github.com/objectiv/objectiv-analytics/pull/852/commits/10397788d40f0ef62c196d0f1bf6e8e049624847

File: bach/tests/functional/bach/test_series_json.py (hunk 1 above)
Comment (KathiaBarahona) [reply bach/tests/functional/bach/test_series_json.py:266]: Mmmm if you added the correct pydoc, it seems that's a seperate issue. Regards, the inclusive issue.. seems that modelhub uses only dicts for start slice, which I only found in tests and here https://github.com/objectiv/objectiv-analytics/blob/4a6cd2479387512cf52fc89630b56bb54da9e228/modelhub/modelhub/series/series_objectiv.py#L171

Comment (hendrik-obj) [reply bach/tests/functional/bach/test_series_json.py:266]: sorry for late to this party. 
Yes, the inspiration for doing it as it was (including the value) was that it is the same as `.loc`. 
Another reason (or maybe the reason for the loc implementation) is that it is in many cases, easier. For example, it is easy to get from the slice that includes the matching end-dict, to the matching dict minus the last object. The other way around however, is difficult.

Comment (hendrik-obj) [reply bach/tests/functional/bach/test_series_json.py:266]: So, it is not 'broken' as https://github.com/objectiv/objectiv-analytics/issues/853 suggests.

File: bach/bach/series/series_json.py [hunk 3]
```diff
@@ -465,6 +474,37 @@ def _get_slice_partial_expr(self, value: Optional[int], is_start: bool) -> Expre
...
+            expression=Expression.construct('element'),
+        )
+
+        all_filters = [
+            element_json.json.get_value(filter_key, as_str=True) == str(filter_value)
+            for filter_key, filter_value in filtering_slice.items()
+        ]
+        slicing_mask = reduce(operator.and_, all_filters)
+        fmt = (
+            f"(select {'min' if is_start else 'max'}(case when {{}} then pos end) "
```
Comment (hendrik-obj) [inline bach/bach/series/series_json.py:503]:
```suggestion
            f"(select {'min' if is_start else '1 + max'}(case when {{}} then pos end) "
```
not fully tested, but something like this gives the expected `.loc`-like behaviour

**Review by hendrik-obj — CHANGES_REQUESTED** [review]
imo we should keep the postgres behaviour. see comments.

File: bach/tests/functional/bach/test_series_json.py (hunk 1 above)
Comment (thijs-obj) [reply bach/tests/functional/bach/test_series_json.py:266]: I don't know `.loc` well. How does that handle 'normal' slices with integers? Does that handle it in the same way we do, or should we change that too?

Comment (hendrik-obj) [reply bach/tests/functional/bach/test_series_json.py:266]: But seeing the discussion, there might be something that is more clear for everybody? finding the position of a matching object in an array as a separate function for example? downside is that the syntax will get quite big if you then want to do a slices.

**Comment by hendrik-obj** [comment]
this test should pass in this pr for bq now: `test_objectiv_stack_type3` in `test_series_objectiv.py`. good to keep track of this somehow.

**Pull request merged**